crates/wasmtime/src/runtime/types.rs

```diff
@@ -867,7 +875,12 @@ impl HeapType {
             (HeapType::NoFunc, _) => false,
 
             (HeapType::ConcreteFunc(_), HeapType::Func) => true,
-            (HeapType::ConcreteFunc(a), HeapType::ConcreteFunc(b)) => a.matches(b),
+            (HeapType::ConcreteFunc(a), HeapType::ConcreteFunc(b)) => {
+                assert!(a.comes_from_same_engine(b.engine()));
+                a.engine()
+                    .signatures()
+                    .is_subtype(a.type_index(), b.type_index())
+            }
```

Comment:
Aha this clears up some historical confusion on my part! This is the missing link in the "break the recursion" thinking for me.

Reply:
Also, for the arms modified here, should the `matches` method be removed? Or is it used elsewhere?

Reply:
They are used elsewhere (e.g. inside the `with_finality_and_supertype` constructors) and are also generally useful as part of the public API.